examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -99,98 +124,210 @@ Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWra
+    {
+    case FanControl::Attributes::PercentSetting::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  PercentSetting");
+        DataModel::Nullable<Percent> percentSetting =
+            NumericAttributeTraits<DataModel::Nullable<Percent>>::StorageToWorking(*value);
+
+        // The cluster code in fan-control-server.cpp is the only one allowed to set PercentSetting to null.
+        // This happens as a consequence of setting the FanMode to kAuto. In auto mode, percentCurrent should continue to report the
+        // real fan speed percentage. In this example, we set PercentCurrent to 0 here as we don't have a real value for the Fan
+        // speed or a FanAutoMode simulator.
+        // When not Null, SpeedCurrent tracks SpeedSetting's value.
+        SetPercentCurrent(percentSetting.ValueOr(0));
+        break;
+    }
+
+    case FanControl::Attributes::SpeedSetting::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  SpeedSetting");
+
+        DataModel::Nullable<uint8_t> speedSetting = NumericAttributeTraits<DataModel::Nullable<uint8_t>>::StorageToWorking(*value);
+        // The cluster code in fan-control-server.cpp is the only one allowed to set speedSetting to null.
+        // This happens as a consequence of setting the FanMode to kAuto. In auto mode, speedCurrent should continue to report the
+        // real fan speed. In this example, we set SpeedCurrent to 0 here as we don't have a real value for the Fan speed or a
+        // FanAutoMode simulator.
+        // When not Null, SpeedCurrent tracks SpeedSetting's value.
+        SetSpeedCurrent(speedSetting.ValueOr(0));
+        // Determine if the speed change should also change the fan mode
+        FanControl::Attributes::FanMode::Set(mEndpoint, SpeedToFanMode(mSpeedCurrent));
+        break;
+    }
+
+    case FanControl::Attributes::FanMode::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  FanMode");
+
+        FanControl::FanModeEnum fanMode = static_cast<FanControl::FanModeEnum>(*value);
+        FanModeWriteCallback(fanMode);
+        break;
+    }
+
+    default: {
+        break;
+    }
+    }
+}
 
-    switch (aPath.mAttributeId)
+FanControl::FanModeEnum ChefFanControlManager::SpeedToFanMode(uint8_t speed)
+{
+    if (speed == 0)
+    {
+        return FanControl::FanModeEnum::kOff;
+    }
+    if (kFanModeLowSpeedRange.Contains(speed))
+    {
+        return FanControl::FanModeEnum::kLow;
+    }
+    if (kFanModeMediumSpeedRange.Contains(speed))
     {
-    case SpeedSetting::Id: {
-        Nullable<uint8_t> newSpeedSetting;
-        ReturnErrorOnFailure(aDecoder.Decode(newSpeedSetting));
+        return FanControl::FanModeEnum::kMedium;
+    }
+    if (kFanModeHighSpeedRange.Contains(speed))
+    {
+        return FanControl::FanModeEnum::kHigh;
+    }
+    return FanControl::FanModeEnum::kUnknownEnumValue;
```

Comment:
Might be OK depending on how this is used, but I would be _really_ careful about ever returning kUnknownEnumValue.  It must never be encoded on the wire, for example, or written into the data store, and this is very much returning a value that will be written into the data store.

I'd suggest a followup restructuring this to make sure that we never write kUnknownEnumValue into the data store.